File: extensions/theme/material.cpp
#include "material.h"

#include "color.h"

using material_color_utilities::Hct;

namespace MaterialColors {

const std::string primary = "#00bfff";
const std::string error = "#ff0062";

bool isBlue(double hue) { return hue >= 200 && hue <= 260; }

double getNeutralChroma(double hue) { return isBlue(hue) ? 8.0 : 2.0; }

std::string hctToHex(double hue, double chroma, double tone) {
  Hct hct(hue, chroma, tone);
  return Color::hexFromArgb(hct.ToInt());
}

Hct hexToHct(const std::string& hex) { return Hct(Color::argbFromHex(hex)); }

ColorMap createPrimaryVariants(const Hct& sourceColor, bool dark) {
  double hue = sourceColor.get_hue();
  double chroma = 40.0;

  return {{"color", hctToHex(hue, chroma, dark ? 80 : 40)},
          {"foreground", hctToHex(hue, chroma, dark ? 20 : 98)}};
}

ColorMap createSecondaryVariants(const Hct& sourceColor, bool dark,
                                 double chroma) {
  double hue = sourceColor.get_hue();
  if (chroma == 0) chroma = dark ? 26.0 : 32.0;

  return {{"color", hctToHex(hue, chroma, dark ? 30 : 80)},
          {"foreground", hctToHex(hue, chroma, dark ? 80 : 20)}};
}

std::string createBorderColor(const Hct& sourceColor, bool dark) {
  double hue = sourceColor.get_hue();
  double chroma = getNeutralChroma(hue);
  return hctToHex(hue, chroma, dark ? 20 : 80);
}

ColorMap createSurfaceVariants(const Hct& sourceColor, bool dark,
                               double chroma) {
  double hue = sourceColor.get_hue();
  if (chroma == 0) chroma = getNeutralChroma(hue);

  return {{"1", hctToHex(hue, chroma, dark ? 8 : 99)},
          {"2", hctToHex(hue, chroma * 1.2, dark ? 12 : 95)},
          {"3", hctToHex(hue, chroma * 1.3, dark ? 13 : 94)},
          {"4", hctToHex(hue, chroma * 1.4, dark ? 15 : 92)}};
}

ColorMap createTextVariants(const Hct& sourceColor, bool dark) {
  double hue = sourceColor.get_hue();
  double chroma = getNeutralChroma(hue);
  return {{"foreground", hctToHex(hue, chroma, dark ? 85 : 15)},
          {"mutedForeground", hctToHex(hue, chroma, dark ? 60 : 45)}};
}

DynamicPalette createDynamicPalette(const Hct& sourceColor, bool dark) {
  auto primary = createPrimaryVariants(sourceColor, dark);
  auto secondary = createSecondaryVariants(sourceColor, dark);
  auto surfaces = createSurfaceVariants(sourceColor, dark);
  auto text = createTextVariants(sourceColor, dark);

  DynamicPalette palette;
  palette.foreground = text.at("foreground");
  palette.mutedForeground = text.at("mutedForeground");
  palette.background = surfaces.at("1");
  palette.card = surfaces.at("2");
  palette.popover = surfaces.at("3");
  palette.hover = surfaces.at("4");
  palette.primary = primary.at("color");
  palette.primaryForeground = primary.at("foreground");
  palette.secondary = secondary.at("color");
  palette.secondaryForeground = secondary.at("foreground");
  palette.border = createBorderColor(sourceColor, dark);

  return palette;
}

}

File: zig/theme/material.zig
const hctMod = @import("material-color-utilities/hct.zig");
const color = @import("color.zig");

const Hct = hctMod.Hct;

pub const defaultColor = "#00bfff";

pub const DynamicPalette = struct {
    foreground: [7]u8,
    mutedForeground: [7]u8,
    background: [7]u8,
    card: [7]u8,
    popover: [7]u8,
    hover: [7]u8,
    primary: [7]u8,
    primaryForeground: [7]u8,
    secondary: [7]u8,
    secondaryForeground: [7]u8,
    border: [7]u8,
};

fn isBlue(hue: f64) bool {
    return hue >= 200 and hue <= 260;
}

fn getNeutralChroma(hue: f64) f64 {
    return if (isBlue(hue)) 8.0 else 2.0;
}

fn hctToHex(hue: f64, chroma: f64, tone: f64) [7]u8 {
    const h = Hct.fromHct(hue, chroma, tone);
    return color.hexFromArgb(h.toInt());
}

pub fn createDynamicPalette(sourceHue: f64, sourceChroma: f64, dark: bool) DynamicPalette {
    _ = sourceChroma;
    const hue = sourceHue;
    const neutralChroma = getNeutralChroma(hue);
    const primaryChroma = 40.0;
    const secondaryChroma: f64 = if (dark) 26.0 else 32.0;

    return .{
        .foreground = hctToHex(hue, neutralChroma, if (dark) 85 else 15),
        .mutedForeground = hctToHex(hue, neutralChroma, if (dark) 60 else 45),
        .background = hctToHex(hue, neutralChroma, if (dark) 8 else 99),
        .card = hctToHex(hue, neutralChroma * 1.2, if (dark) 12 else 95),
        .popover = hctToHex(hue, neutralChroma * 1.3, if (dark) 13 else 94),
        .hover = hctToHex(hue, neutralChroma * 1.4, if (dark) 15 else 92),
        .primary = hctToHex(hue, primaryChroma, if (dark) 80 else 40),
        .primaryForeground = hctToHex(hue, primaryChroma, if (dark) 20 else 98),
        .secondary = hctToHex(hue, secondaryChroma, if (dark) 30 else 80),
        .secondaryForeground = hctToHex(hue, secondaryChroma, if (dark) 80 else 20),
        .border = hctToHex(hue, neutralChroma, if (dark) 20 else 80),
    };
}

pub fn hexToHct(hex: []const u8) Hct {
    return Hct.fromArgb(color.argbFromHex(hex));
}
